# <img style="float: left; padding-right: 10px; width: 45px" src="https://github.com/Harvard-IACS/2021-s109a/blob/master/lectures/crest.png?raw=true"> CS-S109A Introduction to Data Science 

## Final Exam: COVID-19 Modeling

**Harvard University**<br/>
**Summer 2021**<br/>
**Instructors**: Kevin Rader


<hr style='height:2px'>

---



### INSTRUCTIONS

- This final exam is to be completed indivudally.  Do not consult with your peers when working on it (you can aks the teaching staff for clarification questions, including private messages on Ed).
- To submit your assignment follow the instructions given in Canvas.
- Restart the kernel and run the whole notebook again before you submit. 
- As much as possible, try and stick to the hints and functions we import at the top of the homework, as those are the ideas and tools the class supports and is aiming to teach. And if a problem specifies a particular library you're required to use that library, and possibly others from the import list.
- Please use .head() when viewing data. Do not submit a notebook that is excessively long because output was not suppressed or otherwise limited. 

**Note: for all problems, it is up to you to decide how to transform the data (standardization, log transformations, etc.).  Be sure you use and interpret theses transformations approporiately.**

In [1]:
import numpy as np
import pandas as pd
import scipy as sp
import sklearn as sk
import statsmodels as sm
import matplotlib.pyplot as plt
import seaborn as sns

# You are free to use any functions/methods within these packages (BS4, ELI5, and LIME are fine too)
# if you would like to use any other, please contact hte teaching staff 

<hr style="height:2pt">

# Analyzing the recent spread of COVID-19 

![](fig/vaccine.jpeg)

You are tasked with using the COVID case and vaccination data across counties presented by the CDC to analyze the recent surge in COVID infections and the association with (amonth other predictors).  You are also tasked with building prediction models to forecast how the disease spread will change based on data from the previous week (and  demographic and other measures.

The exam broken into 4 problems:
- Problem 1: Data Wrangling and Explorations
- Problem 2: Interpretive Linear Regression Modeling
- Problem 3: Prediction Modeling
- Problem 4: Further Analysis

You are provided with four raw data files, and a 5th cleaned file is provided to be used for all EDA and modeling tasks.

The variables included in each of the four raw data sets are:

For 'covid_cases_county.csv' (note: counties show up many times in this dataset: once for each data they report the number of cases):
- `date`: the date of the measurement, taken weekly
- `county`: county name
- `state`: the state in which the county lies
- `fips`: the unique Federal Information Processing System (FIPS) codes for the county
- `cases`: the cumulative number of confirmed positive cases up to and including that date
- `deaths`: the cumulative number of confirmed COVID-related deaths up to and including that date


For 'vaccines_county.csv' (note: counties show up many times in this dataset: once for each data they report the number of cases):
- `date`: the date of the measurement, taken weekly
- `fips`: the unique FIPS code for the county
- `fully`: the percent of residents that are fully vaccinated in the county on that date
- `dose1`: the percent of residents that have received at least one vaccine dose in the county on that date.

For 'masks_county.csv' (note: this is based on a survey conducted by the New York Times in summer of 2020):
- `fips`: the unique FIPS code for the county
- `never`: the percent of respondents that report they never wore masks in public
- `rarely`: the percent of respondents that report they rarely wore masks in public
- `sometimes`: the percent of respondents that report they sometimes wore masks in public	
- `frequently`: the percent of respondents that report they frequently wore masks in public	
- `always`: the percent of respondents that report they always wore masks in public

For 'demographics_county.csv' (note: these are various measures taken from 2010 to 2020):
- `fips`: the unique FIPS code for the county
- `population`: total number of residents in the country	
- `hispanic`: the percentage of residents that self-identify as hispanic
- `minority`: the percentage of residents that self-identify as a minority group (non-white)
- `female`: the percentage of residents that self-identify as female
- `unemployed`: the percentage of residents that are unemployed
- `income`: the median household income, in thousnads of dollards
- `nodegree`: the percentage of residents that report not having graduated high school
- `bachelor`: the percentage of residents that report having a college degree
- `inactivity`: the percentage of residents that get less than 1 hour of vigorous exercise a week
- `obesity`: the percentage of residents that are considered obese based on BMI
- `density`: the population density (residents per square mile)
- `votergap20`: Biden voting percentage minus Trump voting percentage in the 2020 election
- `votergap16`: Clinton voting percentage minus Trump voting percentage in the 2016 election


### Data Sources
- Vaccinations [here](https://data.cdc.gov/Vaccinations/COVID-19-Vaccinations-in-the-United-States-County/8xkx-amqh).
- Cases [here](https://github.com/nytimes/covid-19-data).
- Mask Usage [here](https://github.com/nytimes/covid-19-data/tree/master/mask-use).
- Demographics [here](https://www.ers.usda.gov/data-products/county-level-data-sets/) 
- 2020 Election [here](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/VOQCHQ)



## Question 1 [25pts]: Data Wrangling and Explorations </b></div>

**1.1** Load the data sets as follows:
- 'covid_cases_county.csv' as `covid_raw` 
- 'vaccines_county.csv' as `vaccines_raw`
- 'masks_county.csv' as `masks`
- 'demographics_county.csv' as `demo` 

**1.2** Create a subset of the `covid_raw` data frame that only contains the measures for 5 dates: June 27 and July 4, 11, 18 and 25.  Do the same for the `vaccines_raw`.  Call these subsets `covid` and `vaccines`, respectively, and print out their dimensions (aka, shapes).

**1.3** Determine and print the number of counties that are measured for each time period in `covid` and `vaccines` (do not print out the list of counties, just the number/count).  Comment on what this implies for presence of missing data.

**1.4** Process both `covid` and `vaccines` so that each county is represented by a single row in each data frame (rather than having 5 separate rows for each county: 1 for each time period in part 1.2).  Call these new generate Pandas data frames `covid_by_county` and `vaccines_by_county` separately.  Print out the dimensions of each resulting data frame, and view the header of `covid_by_county`.  Note: you should use informative names for the columns in the resulting data frames: for example, `cases_w30` for the cumulative number of cases on July 25 (it's the 30th week of the calendar year).

**Hint**: Splitting based on dates and then using `pd.DataFrame.merge` (source)[https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html] could be helpful for this task using the `fips` code as the keys to join on (you should drop any counties that are not measured in all time periods...the default argument for `how` in `pd.DataFrame.merge` will behave this way).

**1.5** Merge the 4 data fames (`covid_by_county`, `vaccines_by_county`, `masks`, and `demo`) based on `fips` and save the result as `covid_merged` (you should drop any counties that are not measured in all 4 data frames).  Determine and report how many counties were dropped from `demo` in this process, and view the header of `covid_merged`.

**1.6** Use `covid_merged` to calculate the novel case rate (per 1000 residents) for each of the weeks for all of the counties, and save these as 4 new well-named variables in `covid_merged`.  For example, `rate_w30` can mathematically be represented as `1000*(cases_30-cases_29)/population`.  Plot the histogram of the novel case rate in week 29, `rate_w29`, and comment on what you notice.

**1.7** We did the steps above (and some other minimal processing) and saved the results in `covid_clean.csv` for you.  Use this data file to answer some exploratory questions and all future analyses: 

1. Has the overall average case rate increased from week 28 (July 5-11) to week 29 (July 12-18)?  
2. Treating the counties as separate and equal observations: in what states did the case rate increase the most?  In what states did the case rate decrease the most (or increse the least)?  List the top 5 for each.  Do you notice any patterns in these states?
3. Create and interpret separate visuals to display how the country case rate in week 29 relates to each of the following variables. Interpret what you see (be specific to this domain).

    a. The political views in the county (as measured by the votergap in the 2020 election).
    
    b. The vaccination rate in the county (for week 28) (be sure to throw away the zeros as these represent unreported values).
    
    c. The population density of the county.
    
    d. Whether 50% or more of the surveyed residents in the county report that they always wore a mask in public at the time of the survey.

## Answers

**1.1** Load the data sets as follows:
- 'covid_cases_county.csv' as `covid_raw` 
- 'vaccines_county.csv' as `vaccines_raw`
- 'masks_county.csv' as `masks`
- 'demographics_county.csv' as `demo` 

Print out each of their dimensions (aka, shapes).

In [134]:
# Saving data sets
covid_raw = pd.read_csv('data/covid_cases_county.csv')
vaccines_raw = pd.read_csv('data/vaccines_county.csv')
masks = pd.read_csv('data/masks_county.csv')
demo = pd.read_csv('data/demographics_county.csv')

# Printing dimentions
print(f'The shape of covid_raw is {covid_raw.shape}')
print(f'The shape of vaccines_raw is {vaccines_raw.shape}')
print(f'The shape of masks is {masks.shape}')
print(f'The shape of demo is {demo.shape}')

The shape of covid_raw is (97394, 6)
The shape of vaccines_raw is (96720, 4)
The shape of masks is (3142, 6)
The shape of demo is (3114, 14)


**1.2** Create a subset of the `covid_raw` data frame that only contains the measures for 5 dates: June 27 and July 4, 11, 18 and 25.  Do the same for the `vaccines_raw`.  Call these subsets `covid` and `vaccines`, respectively, and print out their dimensions (aka, shapes).


In [3]:
# Creating a subset for specified dates:
covid = covid_raw[(covid_raw['date']=='2021-06-27') | 
                  (covid_raw['date']=='2021-07-04') | 
                  (covid_raw['date']=='2021-07-11') |
                  (covid_raw['date']=='2021-07-18') | 
                  (covid_raw['date']=='2021-07-25')]

In [152]:
# Doing the same for vaccines_raw:
vaccines = vaccines_raw[(vaccines_raw['date']=='2021-06-27') | 
                  (vaccines_raw['date']=='2021-07-04') | 
                  (vaccines_raw['date']=='2021-07-11') |
                  (vaccines_raw['date']=='2021-07-18') | 
                  (vaccines_raw['date']=='2021-07-25')]

# Printing dimentions
print(f'The shape of covid is {covid.shape}')
print(f'The shape of vaccines is {vaccines.shape}')
# vaccines

The shape of covid is (16227, 6)
The shape of vaccines is (16120, 4)


**1.3** Determine and print the number of counties that are measured for each time period in `covid` and `vaccines` (do not print out the list of counties, just the number/count).  Comment on what this implies for presence of missing data.


In [174]:


list(np.unique(covid_by_county.fips))[-1000:]
# len(covid[covid['date']=='2021-07-18'].fips)

[40033.0,
 40035.0,
 40037.0,
 40039.0,
 40041.0,
 40043.0,
 40045.0,
 40047.0,
 40049.0,
 40051.0,
 40053.0,
 40055.0,
 40057.0,
 40059.0,
 40061.0,
 40063.0,
 40065.0,
 40067.0,
 40069.0,
 40071.0,
 40073.0,
 40075.0,
 40077.0,
 40079.0,
 40081.0,
 40083.0,
 40085.0,
 40087.0,
 40089.0,
 40091.0,
 40093.0,
 40095.0,
 40097.0,
 40099.0,
 40101.0,
 40103.0,
 40105.0,
 40107.0,
 40109.0,
 40111.0,
 40113.0,
 40115.0,
 40117.0,
 40119.0,
 40121.0,
 40123.0,
 40125.0,
 40127.0,
 40129.0,
 40131.0,
 40133.0,
 40135.0,
 40137.0,
 40139.0,
 40141.0,
 40143.0,
 40145.0,
 40147.0,
 40149.0,
 40151.0,
 40153.0,
 41001.0,
 41003.0,
 41005.0,
 41007.0,
 41009.0,
 41011.0,
 41013.0,
 41015.0,
 41017.0,
 41019.0,
 41021.0,
 41023.0,
 41025.0,
 41027.0,
 41029.0,
 41031.0,
 41033.0,
 41035.0,
 41037.0,
 41039.0,
 41041.0,
 41043.0,
 41045.0,
 41047.0,
 41049.0,
 41051.0,
 41053.0,
 41055.0,
 41057.0,
 41059.0,
 41061.0,
 41063.0,
 41065.0,
 41067.0,
 41069.0,
 41071.0,
 42001.0,
 42003.0,
 42005.0,


date         county          state  fips   cases   deaths
90992  2021-07-18        Unknown         Alaska   NaN     313      0.0
91082  2021-07-18        Unknown       Arkansas   NaN    3093      0.0
91217  2021-07-18        Unknown    Connecticut   NaN    1176      1.0
91222  2021-07-18        Unknown       Delaware   NaN     227      0.0
91287  2021-07-18        Unknown        Florida   NaN       3   1422.0
91436  2021-07-18        Unknown        Georgia   NaN    3036      6.0
91452  2021-07-18        Unknown           Guam   NaN    9446    143.0
91593  2021-07-18        Unknown       Illinois   NaN       0    440.0
91785  2021-07-18        Unknown           Iowa   NaN     455      0.0
92078  2021-07-18        Unknown      Louisiana   NaN     394      0.0
92124  2021-07-18        Unknown       Maryland   NaN       0     16.0
92141  2021-07-18        Unknown  Massachusetts   NaN   47434     12.0
92222  2021-07-18        Unknown       Michigan   NaN     749     16.0
92305  2021-07-18        Unknown      Minnesota   NaN     575     93.0
92448  2021-07-18         Joplin       Missouri   NaN    7224    117.0
92449  2021-07-18    Kansas City       Missouri   NaN   47015    609.0
92657  2021-07-18        Unknown       Nebraska   NaN      66      0.0
92691  2021-07-18        Unknown  New Hampshire   NaN     307      1.0
92712  2021-07-18        Unknown     New Jersey   NaN     826      0.0
92775  2021-07-18  New York City       New York   NaN  962484  33484.0
92954  2021-07-18        Unknown   North Dakota   NaN      31     31.0
93041  2021-07-18        Unknown           Ohio   NaN       0      3.0
93122  2021-07-18        Unknown       Oklahoma   NaN    2722      0.0
93302  2021-07-18        Unknown    Puerto Rico   NaN    5630   2561.0
93314  2021-07-18        Unknown   Rhode Island   NaN   12023     10.0
93515  2021-07-18        Unknown      Tennessee   NaN    7594     95.0
93802  2021-07-18        Unknown           Utah   NaN    1225     27.0
93819  2021-07-18        Unknown        Vermont   NaN      20      0.0

In [161]:
# There are many cointies names that appears in several states. Therefore we need to use the fips column to identify the 
#number of unique counties for both sets.

# Printing the number of counties for each date in covid, incase one date has less counties:
dates = ['2021-06-27','2021-07-04','2021-07-11','2021-07-18','2021-07-25']
print('The number of counties in the covid set is:')
for d in dates:
    nrCounties = len(covid.fips[covid['date']==d].value_counts())  
    print(f' {nrCounties} at date {d}')
# Choosing not to remove null values yet, and using value_counts to get the number of real values for fips.

The number of counties in the covid set is:
 3218 at date 2021-06-27
 3218 at date 2021-07-04
 3218 at date 2021-07-11
 3218 at date 2021-07-18
 3218 at date 2021-07-25


In [122]:
# The number of counties in vaccines:
print('The number of counties in the vaccines set is:')
for d in dates:
    nrFips = len(vaccines.fips[vaccines['date']==d].value_counts())
    print(f' {nrFips} at date {d}')

The number of counties in the vaccines set is:
 3224 at date 2021-06-27
 3224 at date 2021-07-04
 3224 at date 2021-07-11
 3224 at date 2021-07-18
 3224 at date 2021-07-25


##### Answer 1.3:
The vaccines set has 3224 counties while covid has 3218.
This is because the data sets are also including counties in Puerto Rico, Virgin Islands, Northern Mariana and Guam. If you were to only count the 50 states of US you would get 3218-83= 3135 counties for the covid set and 3224-82= 3142 counties in the vaccines set. The number for the vaccines set is almost the same as wikipedias number of counties for the 50 states of US (3143 in 2020), while the covid set is missing 7 more counties. Three of the missing counties have the state column filled, and can be identified as York City, Cansas City and Joplin. These should be added in if we want to use this for prediction or EDA, especially New York City because of its large population.

**1.4** Process both `covid` and `vaccines` so that each county is represented by a single row in each data frame (rather than having 5 separate rows for each county: 1 for each time period in part 1.2).  Call these new generate Pandas data frames `covid_by_county` and `vaccines_by_county` separately.  Print out the dimensions of each resulting data frame, and view the header of `covid_by_county`.  Note: you should use informative names for the columns in the resulting data frames: for example, `cases_w30` for the cumulative number of cases on July 25 (it's the 30th week of the calendar year).

**Hint**: Splitting based on dates and then using `pd.DataFrame.merge` (source)[https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html] could be helpful for this task using the `fips` code as the keys to join on (you should drop any counties that are not measured in all time periods...the default argument for `how` in `pd.DataFrame.merge` will behave this way).




In [120]:
# Splitting the dataset into separate ones for each week:
w26 = vaccines[vaccines['date']=='2021-06-27'].dropna()
w27 = vaccines[vaccines['date']=='2021-07-04'].dropna()
w28 = vaccines[vaccines['date']=='2021-07-11'].dropna()
w29 = vaccines[vaccines['date']=='2021-07-18'].dropna()
w30 = vaccines[vaccines['date']=='2021-07-25'].dropna()
   # Forsøk på å gjøre mergen med for loop. Delete:
# listV = [w30, w29, w28, w27, w26]
# newd = pd.DataFrame(0, index=np.arange(len(w30)),columns=i in range())#, columns='ert')
# for i, dataframe in range(1):#enumerate(liste):
#     if i != 4:
#         newd[:,1+i:1+i*3] = dataframe.merge(liste[i+1], on='fips', how='inner')
#     else:
#         break

mergedV = w26.merge(w27.merge(w28.merge(w29.merge(w30,on='fips'),on='fips'),on='fips'),on='fips')
mergedV.pop('date')
mergedV.pop('date_x')
mergedV.pop('date_y')
vaccines_by_county=mergedV.rename(columns={'fully'  :'fully_w26','dose1'  :'dose1_w26',
                                           'fully_x':'fully_w27','dose1_x':'dose1_w27',
                                           'fully_y':'fully_w28','dose1_y':'dose1_w28',
                                           'fully_x':'fully_w29','dose1_x':'dose1_w29',
                                           'fully_y':'fully_w30','dose1_y':'dose1_w30'})

print(f'The shape of vaccines_by_county is {vaccines_by_county.shape}')
# vaccines_by_county.tail(50)

The shape of vaccines_by_county is (3223, 11)


In [168]:
# Same for the covid set:
c26 = covid[covid['date']=='2021-06-27'].dropna()
c27 = covid[covid['date']=='2021-07-04'].dropna()
c28 = covid[covid['date']=='2021-07-11'].dropna()
c29 = covid[covid['date']=='2021-07-18'].dropna()
c30 = covid[covid['date']=='2021-07-25'].dropna()

mergedC = c26.merge(c27.merge(c28.merge(c29.merge(c30,on='fips'),on='fips'),on='fips'),on='fips')
mergedC.pop('date')
mergedC.pop('date_x')
mergedC.pop('date_y')
mergedC.pop('state_x')
mergedC.pop('state_y')
mergedC.pop('county_x')
mergedC.pop('county_y')

covid_by_county=mergedC.rename(columns={'cases'  :'cases_c26','deaths'  :'deaths_c26',
                                        'cases_x':'cases_c27','deaths_x':'deaths_c27',
                                        'cases_y':'cases_c28','deaths_y':'deaths_c28',
                                        'cases_x':'cases_c29','deaths_x':'deaths_c29',
                                        'cases_y':'cases_c30','deaths_y':'deaths_c30'})

print(f'The shape of vaccines_by_county is {vaccines_by_county.shape}')
print(f'The shape of covid_by_county is {covid_by_county.shape}')
covid_by_county.head(0)

The shape of vaccines_by_county is (3223, 11)
The shape of covid_by_county is (3140, 13)


Empty DataFrame
Columns: [county, state, fips, cases_c26, deaths_c26, cases_c29, deaths_c29, cases_c30, deaths_c30, cases_c29, deaths_c29, cases_c30, deaths_c30]
Index: []

##### Answer 1.4:
Some counties, including all the counties from Puerto Rico and Guam has been dropped from the covid_by_county set because there were missing values, while the vaccines_by_county kept most counties.

**1.5** Merge the 4 data fames (`covid_by_county`, `vaccines_by_county`, `masks`, and `demo`) based on `fips` and save the result as `covid_merged` (you should drop any counties that are not measured in all 4 data frames).  Determine and report how many counties were dropped from `demo` in this process, and view the header of `covid_merged`.



**1.6** Use `covid_merged` to calculate the novel case rate (per 1000 residents) for each of the weeks for all of the counties, and save these as 4 new well-named variables in `covid_merged`.  For example, `rate_w30` can mathematically be represented as `1000*(cases_30-cases_29)/population`.  Plot the histogram of the novel case rate in week 29, July 12-18, `rate_w29`, and comment on what you notice.



In [ ]:
######
# your code here
######

*your answer here*

**1.7** We did the steps above (and some other minimal processing) and saved the results in `covid_clean.csv` for you.  Use this data file to answer some exploratory questions and all future analyses: 

1. Has the overall average case rate increased from week 28 (July 5-11) to week 29 (July 12-18)?  
2. Treating the counties as separate and equal observations: in what states did the case rate increase the most?  In what states did the case rate decrease the most (or increse the least)?  List the top 5 for each.  Do you notice any patterns in these states?
3. Create and interpret separate visuals to display how the country case rate in week 29 relates to each of the following variables. Interpret what you see (be specific to this domain).

    a. The political views in the county (as measured by the votergap in the 2020 election).
    
    b. The vaccination rate in the county (for week 28) (be sure to throw away the zeros as these represent unreported values).
    
    c. The population density of the county.
    
    d. Whether 50% or more of the surveyed residents in the county report that they always wore a mask in public at the time of the survey.

In [ ]:
######
# your code here
#######

Clarifications:
Q1.7.2: this is referring to the case rate increase from week 28 to week 29.
Q1.7.3: you should investigate both forms of vaccination rates.
    
Q1.7 and on: the rate variables in the given covid_clean data set were not multiplied by 1000.   
    Either keeping as is or multiplying by 1000 is fine (coefficient magnitudes are more reasonable if you multiply by 1000).

*your answer here*

---

## Question 2 [35pts]: Regression modeling 

**2.1** Fit a linear regression model to predict `rate_w29` (which represent the rate of new cases in the week of July 12-18) from `rate_w28` (July 5-11). Report the 95% confidence intervals for the coefficients, and carefully interpret the coefficients (including their statistical significances).  What does this model suggest about whether the rate of COVID infection increased from week 28 to week 29?


**2.2** Fit a linear regression model to predict `rate_w29` from `rate_w28` and `votergap20` along with the interaction between the two.  Interpret the coefficient estimates carefully (no need to mention significances).


**2.3** Create a scatterplot of `rate_w29` vs. `rate_w28`.  Add 3 separate predicted lines from your model in 2.2 to this scatterplot: the predicted line from the model in 2.2 for counties...
    1. where Biden was favored by 50 percentage points.
    2. where Biden and Trump were equal
    3. where Trump was favored by 50 percentage points.
Interpret what you see.


**2.4** Fit a linear regression model to assess the overall association of vaccination rate (`fully_w28`) on `rate_w29`.  Carefully interpret the results (including the statistical significance).  


**2.5** Many counties have the value zero for `fully_w28` which really represents a missing/unreported value for vaccinationr rate.  Comment on the effect of ignoring this issue can have on the intepretations and inferences in the model in 2.4.  What would be a better way of handling this issue?


**2.6** What factors could be confounded (whether mesured here or not) with the result seen in the model from 2.3 (list up to 3)?  Fit an appropriate linear model that controls for as many of these factors as possible (for those that are measured in this data set). Interpret the coefficient estimates from this model and compare to the results from 2.4.

**2.7** What major issue could arise if you fit a model to predict `rate_w29` from `rate_w28` and `rate_w27` (or from `fully_w28` and `fully_w27`) in a linear regression model?  Suggest and explain the use of two different approaches to account for this: one approach should be based on modeling and one approach should be based on feature engineering/variable transformations (not PCA). 

**2.8** The test set has a response variable that is `rate_w30`.  How would you use your models to predict `rate_w29` in this section in order to predict `rate_w30` instead?  Explain.  What could go wrong in this modification?

**Hint**: what should be the predictors to predict `rate_w30` instead of `rate_w29`? 


## Answers

**2.1** Fit a linear regression model to predict `rate_w29` (which represent the rate of new cases in the week of July 12-18) from `rate_w28` (July 5-11). Report the 95% confidence intervals for the coefficients, and carefully interpret the coefficients (including their statistical significances).  What does this model suggest about whether the rate of COVID infection increased from week 28 to week 29?


In [ ]:
######
# your code here
######

*your answer here*

**2.2** Fit a linear regression model to predict `rate_w29` from `rate_w28` and `votergap20` along with the interaction between the two.  Interpret the coefficient estimates carefully (no need to mention significances).


In [ ]:
######
# your code here
######

*your answer here*

**2.3** Create a scatterplot of `rate_w29` vs. `rate_w28`.  Add 3 separate predicted lines from your model in 2.2 to this scatterplot: the predicted line from the model in 2.2 for counties...
    1. where Biden was favored by 50 percentage points.
    2. where Biden and Trump were equal
    3. where Trump was favored by 50 percentage points.
Interpret what you see.




In [ ]:
######
# your code here
######

*your answer here*

**2.4** Fit a linear regression model to assess the overall association of vaccination rate (`fully_w28`) on `rate_w29`.  Carefully interpret the results (including the statistical significance).  




In [ ]:
######
# your code here
######

*your answer here*

**2.5** Many counties have the value zero for `fully_w28` which really represents a missing/unreported value for vaccinationr rate.  Comment on the effect of ignoring this issue can have on the intepretations and inferences in the model in 2.4.  What would be a better way of handling this issue?




*your answer here*

**2.6** What factors could be confounded (whether mesured here or not) with the result seen in the model from 2.3 (list up to 3)?  Fit an appropriate linear model that controls for as many of these factors as possible (for those that are measured in this data set). Interpret the coefficient estimates from this model and compare to the results from 2.4.



In [ ]:
######
# your code here
######

*your answer here*

**2.7** What major issue could arise if you fit a model to predict `rate_w29` from `rate_w28` and `rate_w27` (or from `fully_w28` and `fully_w27`) in a linear regression model?  Suggest and explain the use of two different approaches to account for this: one approach should be based on modeling and one approach should be based on feature engineering/variable transformations (not PCA). 



*your answer here*

**2.8** The test set has a response variable that is `rate_w30`.  How would you use your models to predict `rate_w29` in this section in order to predict `rate_w30` instead?  Explain.  What could go wrong in this modification?

**Hint**: what should be the predictors to predict `rate_w30` instead of `rate_w29`? 


*your answer here*

---

## Question 3 [30pts]: Prediction modeling 

**3.1** Fit a well-tuned lasso model to predict `rate_w29` from the following set of predictors (along with all 2-way interactions among the main effects and all 2nd and 3rd order polynomial terms):

`['rate_w28','rate_w27','dose1_w28','hispanic','minority','female','unemployed', 'income','nodegree','bachelor','inactivity','obesity','density','votergap20']`

Report and explain the best choice of $\lambda$ (a visual can help with this), your estimate of out-of-sample $R^2$, along with the number of coefficients that shrunk exactly to zero (or numerically zero) and the number that are non-zero.

**3.2** Plot the trajectory curves of the main effects `['rate_w28','rate_w27','fully_w28','votergap20']` from this model: the estimates of the $\beta$ coefficients as a function of $\lambda$.  Interpret what you notice.

**3.3** Fit a well-tuned random forest model to predict `rate_w29` from the predictors listed in 3.1.  Report your choice of the tuning parameters and briefly justify your choices (a visual or table may be helpful for this).  Provide an estimate of out-of-sample $R^2$.  Note: do not go to crazy with the number of options for the parameters you are tuning...choose a set of values that are reasonable.

**3.4** Interpret the relationship between `rate_w29` and `dose1_w28` from the random forest model in 3.3.  Is there any evidence of interactive effects in this model involving `dose1_w28`?  How do you know?  Provide a reasonable visual (or a few visuals) to help you with these tasks and interpret what you see. 

**3.5** Fit a well-tuned boosting model to predict `rate_w29` from the predictors listed in 3.1.  Report your best choice of the tuning parameters and briefly justify your choice (a visual or table may be helpful for this).  Provide an estimate of out-of-sample $R^2$.  Note: again, do not go to crazy with the number of options for the parameters you are tuning...choose a set of values that are reasonable.

**3.6** Improve upon your favorite/best predictive model from 3.1, 3.3, or 3.5, by including other provided feature, by doing feature engineering, or by doing variable removal/selection.  Explain your choices.  Provide an estimate of out-of-sample $R^2$. 

**3.7** Evaluate your models from 3.1, 3.3, 3.5, and 3.6 on the test set (this will take some work...refer back to 2.8) using $R^2$.  How do these model's $R^2$ in test compare to the out-of-sample $R^2$ when tuning?  Explain whether this is surprising or not?



## Answers

**3.1** Fit a well-tuned lasso model to predict `rate_w29` from the following set of predictors (along with all 2-way interactions among the main effects and all 2nd and 3rd order polynomial terms):

`['rate_w28','rate_w27','dose1_w28','hispanic','minority','female','unemployed', 'income','nodegree','bachelor','inactivity','obesity','density','votergap20']`

Report and explain the best choice of $\lambda$ (a visual can help with this), your estimate of out-of-sample $R^2$, along with the number of coefficients that shrunk exactly to zero (or numerically zero) and the number that are non-zero.


In [ ]:
######
# your code here
######

Corrections:
Q3.1: please ignore/remove 'cancer' from this list of predictors.
Q3.2: 'fully_w28' should be replaced with 'dose1_w28'

*your answer here*

**3.2** Plot the trajectory curves of the main effects `['rate_w28','rate_w27','fully_w28','votergap20']` from this model: the estimates of the $\beta$ coefficients as a function of $\lambda$.  Interpret what you notice.


In [ ]:
######
# your code here
######

*your answer here*

**3.3** Fit a well-tuned random forest model to predict `rate_w29` from the predictors listed in 3.1.  Report your choice of the best tuning parameters and briefly justify your choice (a visual or table may be helpful for this).  Provide an estimate of out-of-sample $R^2$.  Note: do not go to crazy with the number of options for the parameters you are tuning...choose a set of values that are reasonable.


In [ ]:
######
# your code here
######

*your answer here*

**3.4** Interpret the relationship between `rate_w29` and `dose1_w28` from the random forest model in 3.3.  Is there any evidence of interactive effects in this model involving `dose1_w28`?  How do you know?  Provide a reasonable visual (or a few visuals) to help you with these tasks and interpret what you see. 



In [ ]:
######
# your code here
######

*your answer here*

**3.5** Fit a well-tuned boosting model to predict `rate_w29` from the predictors listed in 3.1.  Report your best choice of the tuning parameters and briefly justify your choice (a visual or table may be helpful for this).  Provide an estimate of out-of-sample $R^2$.  Note: again, do not go to crazy with the number of options for the parameters you are tuning...choose a set of values that are reasonable.



In [ ]:
######
# your code here
######

*your answer here*

**3.6** Improve upon your favorite/best predictive model from 3.1, 3.3, or 3.5, by including other provided feature, by doing feature engineering, or by doing variable removal/selection.  Explain your choices.  Provide an estimate of out-of-sample $R^2$. 



In [ ]:
######
# your code here
######

*your answer here*

**3.7** Evaluate your models from 3.1, 3.3, 3.5, and 3.6 on the test set (this will take some work...refer back to 2.8) using $R^2$.  How do these model's $R^2$ in test compare to the out-of-sample $R^2$ when tuning?  Explain whether this is surprising or not?

In [ ]:
######
# your code here
######

*your answer here*

---

## Question 4 [10pts]: Going further

**4.1** Use all of the useable variables in `demo` and `masks` to create clusters of observations based on the $K$-means clustering approach.  Be sure to carefully select a reasonable choice for $K$.  Explain your choice (a visual may help with this).

**4.2** Use your created clusters and incorporate them as predictor(s) into a linear regression model to assess whether the relationships you measured in the model from 2.6 depend on cluster type.  Comment on what you notice.  Determine whether out-of-sample $R^2$ has improved using this model (in comparison to the model from 2.6) based on 5-fold CV.

**4.3: BONUS** Find data online to improve the prediction accuracy of your best model. Be sure to cite your source of your data and the approach you took into incorporating these new data.  Note: this is only worth up to 3 bonus points, so do not spend too much effor on this part over improving ealrier parts of the exam.

## Answers

**4.1** Use all of the useable variables in `demo` and `masks` to create clusters of observations based on the $K$-means clustering approach.  Be sure to carefully select a reasonable choice for $K$.  Explain your choice (a visual may help with this).

In [ ]:
######
# your code here
######

*your answer here*

**4.2** Use your created clusters and incorporate them as predictor(s) into a linear regression model to assess whether the relationships you measured in the model from 2.6 depend on cluster type.  Comment on what you notice.  Determine whether out-of-sample $R^2$ has improved using this model (in comparison to the model from 2.6) based on 5-fold CV.


In [ ]:
######
# your code here
######

*your answer here*

**4.3: BONUS** Find data online to improve the prediction accuracy of your best model. Be sure to cite your source of your data and the approach you took into incorporating these new data.  Note: this is only worth up to 3 bonus points, so do not spend too much effor on this part over improving ealrier parts of the exam.

In [ ]:
######
# your code here
######



*your answer here*